src/dps.c

```diff
@@ -1166,10 +1175,66 @@ static void NodeRun(void* arg)
     }
 }
 
+static DPS_Status SetCurve(DPS_KeyStoreRequest* request, const DPS_Key* key)
+{
+    int8_t* curve = request->data;
+    uint8_t d[EC_MAX_COORD_LEN];
+
+    switch (key->type) {
+    case DPS_KEY_EC:
+        *curve = key->ec.curve;
+        return DPS_OK;
+    case DPS_KEY_EC_CERT:
+        if (key->cert.privateKey) {
+            return ParsePrivateKey_ECDSA(key->cert.privateKey, key->cert.privateKeyLen,
+                                         key->cert.password, key->cert.passwordLen,
+                                         curve, d);
+        }
+        /* FALLTHROUGH */
+    default:
+        return DPS_ERR_MISSING;
+    }
+}
+
+static DPS_Status GetSignatureAlgorithm(DPS_KeyStore* keyStore, const uint8_t* id, size_t idLen, int8_t* alg)
+{
+    DPS_KeyStoreRequest request;
+    int8_t curve = 0;
+    DPS_Status ret;
+
+    if (!keyStore || !keyStore->keyHandler) {
+        return DPS_ERR_MISSING;
+    }
+
+    memset(&request, 0, sizeof(request));
+    request.keyStore = keyStore;
+    request.data = &curve;
+    request.setKey = SetCurve;
+    ret = keyStore->keyHandler(&request, id, idLen);
+    if (ret != DPS_OK) {
+        return ret;
+    }
+    switch (curve) {
+    case EC_CURVE_P256:
+        *alg = COSE_ALG_ES256;
+        break;
+    case EC_CURVE_P384:
+        *alg = COSE_ALG_ES384;
+        break;
+    case EC_CURVE_P521:
+        *alg = COSE_ALG_ES512;
+        break;
+    default:
+        ret = DPS_ERR_MISSING;
+        break;
+    }
+    return ret;
+}
 
-DPS_Node* DPS_CreateNode(const char* separators, DPS_KeyStore* keyStore, const DPS_UUID* keyId)
+DPS_Node* DPS_CreateNode(const char* separators, DPS_KeyStore* keyStore, const uint8_t* id, size_t idLen)
```

Comment:
As I am reviewing this it seems that might be better to have a DPS_KeyIdentifier (or DPS_SecurityPrincipal - see earlier comment) instead of passing pointer an length around.

Reply:
It sounds like the coupling between the id and the key store is not clear.  A DPS_KeyIdentifier would help, but it would also make the API a bit clumsier.  Instead of

  DPS_CreateNode(..., fooId, sizeof(fooId));

I'd have to do:

  DPS_KeyIdentifier fooId = { value, sizeof(value); };
  DPS_CreateNode(..., &fooId);

But, I'll take a closer look and see how it plays out.  Internally I've already got a bunch of id, idLen members and if the application defined IDs are static then the API shouldn't suffer.